File: frontend/src/services/chunkService.ts
import api, { handleApiError, type ApiError } from './api';
import type {
  LinguisticChunk,
  ChunkMastery,
  CefrLevel,
  ChunkCategory,
} from '../types';

export const chunkService = {
  /**
   * Busca chunks linguísticos por idioma e nível
   */
  getChunks: async (
    languageCode: string,
    cefrLevel?: CefrLevel,
    category?: ChunkCategory
  ): Promise<LinguisticChunk[]> => {
    const params = new URLSearchParams({ languageCode });
    if (cefrLevel) params.append('cefrLevel', cefrLevel);
    if (category) params.append('category', category);
    const response = await api.get<LinguisticChunk[]>(`/chunks?${params}`);
    return response.data;
  },

  /**
   * Busca um chunk específico por ID
   */
  getChunk: async (chunkId: string): Promise<LinguisticChunk> => {
    const response = await api.get<LinguisticChunk>(`/chunks/${chunkId}`);
    return response.data;
  },

  /**
   * Busca chunks essenciais (core) de um nível
   */
  getCoreChunks: async (languageCode: string, cefrLevel: CefrLevel): Promise<LinguisticChunk[]> => {
    const response = await api.get<LinguisticChunk[]>(
      `/chunks/core?languageCode=${languageCode}&cefrLevel=${cefrLevel}`
    );
    return response.data;
  },

  /**
   * Busca domínio do usuário sobre chunks
   */
  getMastery: async (languageCode: string): Promise<ChunkMastery[]> => {
    const response = await api.get<ChunkMastery[]>(`/chunks/mastery?languageCode=${languageCode}`);
    return response.data;
  },

  /**
   * Atualiza domínio de um chunk após prática
   */
  updateMastery: async (
    chunkId: string,
    quality: number,
    context?: string
  ): Promise<ChunkMastery> => {
    const response = await api.post<ChunkMastery>(`/chunks/${chunkId}/practice`, {
      quality,
      context,
    });
    return response.data;
  },

  /**
   * Busca chunks por categoria
   */
  getByCategory: async (
    languageCode: string,
    category: ChunkCategory
  ): Promise<LinguisticChunk[]> => {
    const response = await api.get<LinguisticChunk[]>(
      `/chunks/category/${category}?languageCode=${languageCode}`
    );
    return response.data;
  },

  /**
   * Busca chunks recomendados para praticar (baixo domínio)
   */
  getRecommended: async (languageCode: string, limit: number = 10): Promise<LinguisticChunk[]> => {
    const response = await api.get<LinguisticChunk[]>(
      `/chunks/recommended?languageCode=${languageCode}&limit=${limit}`
    );
    return response.data;
  },

  /**
   * Busca estatísticas de domínio por categoria
   */
  getMasteryStats: async (languageCode: string): Promise<{
    category: ChunkCategory;
    total: number;
    mastered: number;
    avgMasteryLevel: number;
  }[]> => {
    const response = await api.get(`/chunks/mastery/stats?languageCode=${languageCode}`);
    return response.data;
  },
};

export { handleApiError, type ApiError };

File: src/main/java/com/langia/backend/service/ChunkService.java
package com.langia.backend.service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.langia.backend.dto.ChunkMasteryDTO;
import com.langia.backend.dto.LinguisticChunkDTO;
import com.langia.backend.exception.ResourceNotFoundException;
import com.langia.backend.model.ChunkMastery;
import com.langia.backend.model.LinguisticChunk;
import com.langia.backend.model.User;
import com.langia.backend.repository.ChunkMasteryRepository;
import com.langia.backend.repository.LinguisticChunkRepository;
import com.langia.backend.repository.UserRepository;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

/**
 * Service para gerenciamento de chunks linguísticos.
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class ChunkService {

    private final LinguisticChunkRepository chunkRepository;
    private final ChunkMasteryRepository masteryRepository;
    private final UserRepository userRepository;

    /**
     * Busca chunks por idioma e nível.
     */
    public List<LinguisticChunkDTO> getChunks(String languageCode, String cefrLevel, String category) {
        List<LinguisticChunk> chunks;

        if (cefrLevel != null && category != null) {
            chunks = chunkRepository.findByLanguageCodeAndCefrLevelAndCategory(languageCode, cefrLevel, category);
        } else if (cefrLevel != null) {
            chunks = chunkRepository.findByLanguageCodeAndCefrLevel(languageCode, cefrLevel);
        } else if (category != null) {
            chunks = chunkRepository.findByLanguageCodeAndCategory(languageCode, category);
        } else {
            chunks = chunkRepository.findByLanguageCode(languageCode);
        }

        return chunks.stream()
                .map(LinguisticChunkDTO::fromEntity)
                .toList();
    }

    /**
     * Busca chunks essenciais (core) de um nível.
     */
    public List<LinguisticChunkDTO> getCoreChunks(String languageCode, String cefrLevel) {
        return chunkRepository.findCoreChunks(languageCode, cefrLevel)
                .stream()
                .map(LinguisticChunkDTO::fromEntity)
                .toList();
    }

    /**
     * Busca domínio do usuário sobre chunks.
     */
    public List<ChunkMasteryDTO> getMastery(UUID userId, String languageCode) {
        return masteryRepository.findByUserIdAndChunkLanguageCode(userId, languageCode)
                .stream()
                .map(ChunkMasteryDTO::fromEntity)
                .toList();
    }

    /**
     * Atualiza domínio de um chunk após prática.
     */
    @Transactional
    public ChunkMasteryDTO updateMastery(UUID userId, UUID chunkId, int quality, String context) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        LinguisticChunk chunk = chunkRepository.findById(chunkId)
                .orElseThrow(() -> new ResourceNotFoundException("Chunk not found: " + chunkId));

        ChunkMastery mastery = masteryRepository.findByUserIdAndChunkId(userId, chunkId)
                .orElseGet(() -> {
                    ChunkMastery newMastery = new ChunkMastery();
                    newMastery.setUser(user);
                    newMastery.setChunk(chunk);
                    newMastery.setMasteryLevel(0);
                    newMastery.setTimesPracticed(0);
                    newMastery.setTimesCorrect(0);
                    return newMastery;
                });

        // Atualiza nível de domínio baseado na qualidade da resposta
        int newLevel = calculateNewMasteryLevel(mastery.getMasteryLevel(), quality);
        mastery.setMasteryLevel(newLevel);
        mastery.setTimesPracticed(mastery.getTimesPracticed() + 1);
        mastery.setLastPracticedAt(LocalDateTime.now());

        if (context != null && !context.isBlank()) {
            List<String> contexts = mastery.getContextsUsed();
            if (!contexts.contains(context)) {
                contexts.add(context);
                mastery.setContextsUsed(contexts);
            }
        }

        mastery = masteryRepository.save(mastery);

        log.info("User {} practiced chunk {} with quality {}, new mastery level: {}",
                userId, chunkId, quality, newLevel);

        return ChunkMasteryDTO.fromEntity(mastery);
    }

    /**
     * Busca chunks recomendados para praticar (baixo domínio).
     */
    public List<LinguisticChunkDTO> getRecommended(UUID userId, String languageCode, int limit) {
        return masteryRepository.findChunksNeedingPractice(userId, languageCode)
                .stream()
                .limit(limit)
                .map(m -> LinguisticChunkDTO.fromEntity(m.getChunk()))
                .toList();
    }

    private int calculateNewMasteryLevel(int currentLevel, int quality) {
        // quality: 0-5 (como SM-2)
        // mastery: 0-5

        if (quality >= 4) {
            return Math.min(5, currentLevel + 1);
        } else if (quality >= 3) {
            return currentLevel;
        } else if (quality >= 2) {
            return Math.max(0, currentLevel - 1);
        } else {
            return Math.max(0, currentLevel - 2);
        }
    }
}
